xcm/src/v3/multiasset.rs

```diff
@@ -0,0 +1,871 @@
+			_ => Err(()),
+		}
+	}
+}
+
+impl TryFrom<AssetInstance> for u32 {
+	type Error = ();
+	fn try_from(x: AssetInstance) -> Result<Self, ()> {
+		match x {
+			AssetInstance::Index(x) => x.try_into().map_err(|_| ()),
+			_ => Err(()),
+		}
+	}
+}
+
+impl TryFrom<AssetInstance> for u64 {
+	type Error = ();
+	fn try_from(x: AssetInstance) -> Result<Self, ()> {
+		match x {
+			AssetInstance::Index(x) => x.try_into().map_err(|_| ()),
+			_ => Err(()),
+		}
+	}
+}
+
+impl TryFrom<AssetInstance> for u128 {
+	type Error = ();
+	fn try_from(x: AssetInstance) -> Result<Self, ()> {
+		match x {
+			AssetInstance::Index(x) => Ok(x),
+			_ => Err(()),
+		}
+	}
+}
+
+/// Classification of whether an asset is fungible or not, along with a mandatory amount or instance.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub enum Fungibility {
+	Fungible(#[codec(compact)] u128),
+	NonFungible(AssetInstance),
+}
+
+impl Fungibility {
+	pub fn is_kind(&self, w: WildFungibility) -> bool {
+		use Fungibility::*;
+		use WildFungibility::{Fungible as WildFungible, NonFungible as WildNonFungible};
+		matches!((self, w), (Fungible(_), WildFungible) | (NonFungible(_), WildNonFungible))
+	}
+}
+
+impl From<i32> for Fungibility {
+	fn from(amount: i32) -> Fungibility {
+		debug_assert_ne!(amount, 0);
+		Fungibility::Fungible(amount as u128)
+	}
+}
+
+impl From<u128> for Fungibility {
+	fn from(amount: u128) -> Fungibility {
+		debug_assert_ne!(amount, 0);
+		Fungibility::Fungible(amount)
+	}
+}
+
+impl<T: Into<AssetInstance>> From<T> for Fungibility {
+	fn from(instance: T) -> Fungibility {
+		Fungibility::NonFungible(instance.into())
+	}
+}
+
+impl TryFrom<OldFungibility> for Fungibility {
+	type Error = ();
+	fn try_from(value: OldFungibility) -> Result<Self, Self::Error> {
+		use OldFungibility::*;
+		Ok(match value {
+			Fungible(n) => Self::Fungible(n),
+			NonFungible(i) => Self::NonFungible(i.try_into()?),
+		})
+	}
+}
+
+/// Classification of whether an asset is fungible or not.
+#[derive(
+	Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub enum WildFungibility {
+	Fungible,
+	NonFungible,
+}
+
+impl TryFrom<OldWildFungibility> for WildFungibility {
+	type Error = ();
+	fn try_from(value: OldWildFungibility) -> Result<Self, Self::Error> {
+		use OldWildFungibility::*;
+		Ok(match value {
+			Fungible => Self::Fungible,
+			NonFungible => Self::NonFungible,
+		})
+	}
+}
+
+/// Classification of an asset being concrete or abstract.
+#[derive(
+	Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub enum AssetId {
+	Concrete(MultiLocation),
+	Abstract([u8; 32]),
```

Comment:
No change in meaning - it's a named identifier for an asset, rather than a location-based identifier. It is not finding any use at present, and may be a candidate for removal.